src/utils/todoHandlers.ts

```diff
@@ -0,0 +1,238 @@
+  }
+};
+
+export const toggleTodoCompletion = async (
+  id: number,
+  setTodos: React.Dispatch<React.SetStateAction<Todo[]>>,
+  setErrorMessage: React.Dispatch<React.SetStateAction<CustomError>>,
+  setLoadingTodo: React.Dispatch<React.SetStateAction<LoadingTodo>>,
+) => {
+  try {
+    setLoadingTodo({ id, action: 'updating' });
+
+    let changedTodo: Todo | null = null;
+
+    const updateTodos = (todos: Todo[]) =>
+      todos.map(todo => {
+        if (todo.id === id) {
+          changedTodo = { ...todo, completed: !todo.completed };
+
+          return changedTodo;
+        }
+
+        return todo;
+      });
+
+    await setTodos(prevTodos => updateTodos(prevTodos));
+
+    if (changedTodo) {
+      await todoService.changeCompletedStatus(changedTodo);
+    }
+
+    setTimeout(() => setLoadingTodo(null), 900);
+  } catch (error) {
+    setErrorMessage('Unable to update a todo');
+    setLoadingTodo(null);
+  }
+};
+
+export const toggleAllTodosCompletion = async (
+  todos: Todo[],
+  setTodos: React.Dispatch<SetStateAction<Todo[]>>,
+  setErrorMessage: React.Dispatch<SetStateAction<CustomError>>,
+  setLoadingTodos: React.Dispatch<SetStateAction<LoadingTodo[]>>,
+) => {
+  const completedTodoExists = todos.some(todo => todo.completed);
+  const uncompletedTodoExists = todos.some(todo => !todo.completed);
+  const allTodosCompleted = todos.every(todo => todo.completed);
+
+  try {
+    if ((completedTodoExists && uncompletedTodoExists) || !allTodosCompleted) {
+      setTodos(prevTodos =>
+        prevTodos.map(todo => {
+          const updatedTodo = { ...todo, completed: true };
+
+          if (!todo.completed) {
+            setLoadingTodos(prev => [
+              ...prev,
+              { id: updatedTodo.id, action: 'updating' },
+            ]);
+            todoService.changeCompletedStatus(updatedTodo);
+          }
+
+          return updatedTodo;
+        }),
+      );
+
+      setTodos(prevTodos =>
+        prevTodos.map(todo => {
+          return { ...todo, completed: true };
+        }),
+      );
+    } else {
+      setTodos(prevTodos =>
+        prevTodos.map(todo => {
+          const updatedTodo = { ...todo, completed: false };
+
+          setLoadingTodos(prev => [
+            ...prev,
+            { id: updatedTodo.id, action: 'updating' },
+          ]);
+          todoService.changeCompletedStatus(updatedTodo);
+
+          return updatedTodo;
+        }),
+      );
+
+      setTodos(prevTodos =>
+        prevTodos.map(todo => {
+          return { ...todo, completed: false };
+        }),
+      );
+    }
+
+    setTimeout(() => setLoadingTodos([]), 900);
+  } catch (error) {
+    setErrorMessage('Unable to update todos');
+    setLoadingTodos([]);
+  }
+};
+
+export const removeCompletedTodos = async (
+  todos: Todo[],
+  setTodos: React.Dispatch<SetStateAction<Todo[]>>,
+  setErrorMessage: React.Dispatch<SetStateAction<CustomError>>,
+  setLoadingTodos: React.Dispatch<SetStateAction<LoadingTodo[]>>,
```

Comment:
same with functions like this, try to avoid this pattern when you pass setState functions into helper functions, it's really difficult to read